### PR TITLE
[1.16.x] Allow command to be registered for client side execution

### PR DESCRIPTION
**Pull request by GotoLink**
Update of #6670 for 1.16
The old test mod was removed so it is 'new' again.

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package net.minecraftforge.client.event;
+
+import com.mojang.brigadier.CommandDispatcher;
+import net.minecraft.command.ISuggestionProvider;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * Fired when client {@link CommandDispatcher} has all of its commands synced from the server
+ */
+public class CommandRegistryEvent extends Event
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java:29]: As we have a `RegisterCommandsEvent`, it might be good to call this something similar like `RegisterClientCommandsEvent` and have similar context. I would, however, not extend `RegisterCommandsEvent` so that people that listen for `RegisterCommandsEvent` don't accidentally register for the client as well

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java [hunk 2]
```diff
@@ -794,4 +795,52 @@ public static void drawItemLayered(ItemRenderer renderer, IBakedModel modelIn, I
...
+                {
+                    details.func_230529_a_(new StringTextComponent(syntax.getInput().substring(position)).func_240701_a_(TextFormatting.RED, TextFormatting.UNDERLINE));
+                }
+                details.func_230529_a_(new TranslationTextComponent("command.context.here").func_240701_a_(TextFormatting.RED, TextFormatting.ITALIC));
+                Minecraft.getInstance().player.sendMessage(new StringTextComponent("").func_230529_a_(details).func_240699_a_(TextFormatting.RED), Util.field_240973_b_);
+            }
+        }
+        catch (Exception generic)
+        {
+            StringTextComponent message = new StringTextComponent(generic.getMessage() == null ? generic.getClass().getName() : generic.getMessage());
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java:803]: IMO any non-command exception thrown by a client command should be logged

File: patches/minecraft/net/minecraft/command/Commands.java.patch [hunk 3]
```diff
@@ -15,3 +15,11 @@
              return this.field_197062_b.execute(stringreader, p_197059_1_);
           } catch (CommandException commandexception) {
              p_197059_1_.func_197021_a(commandexception.func_197003_a());
+@@ -273,6 +281,7 @@
+                argumentbuilder.executes((p_197053_0_) -> {
+                   return 0;
+                });
++               argumentbuilder.executes(null);// FORGE Client Commands support
```
Comment (ichttt) [inline patches/minecraft/net/minecraft/command/Commands.java.patch]: What does this do? Why is this needed for client commands to work? Please add some more description to the comment

File: src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java (hunk 1 above)
Comment (GotoLink) [reply src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java:29]: The event you refer to has a very different purpose (and was proposed later than this).
I would argue it to be less confusing when the name are not too similar.

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java (hunk 2 above)
Comment (GotoLink) [reply src/main/java/net/minecraftforge/client/ForgeHooksClient.java:803]: This is a paste from the vanilla Commands class error handling.
Behavior is intentionally identical with server commands.

File: patches/minecraft/net/minecraft/command/Commands.java.patch (hunk 3 above)
Comment (GotoLink) [reply patches/minecraft/net/minecraft/command/Commands.java.patch]: This is explained in the commit message:
"Patch Commands to remove empty lambdas of server commands from being
sent to client (and thus execute there when they should not)."
The line right above this is the one where "empty lambdas" are set into the builder and then sent to the client by SCommandListPacket. They will then get registered by CommandDispatcher, and would be trapped by the hook of ForgeHooksClient in ChatScreen.
Without this patch, server commands are executed as client commands which do nothing since empty lambdas are always successful, effectively trapping all server commands on the client side.

File: src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java (hunk 1 above)
Comment (ichttt) [reply src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java:29]: How this this event different in purpose from the RegisterCommandsEvent? Both are intended to be used to register new commands, one for client-side commands, one for server-side commands. There are fired at different times and have a completly different implementation, but to the modder they serve the same purpose for different sides or am I wrong?

File: patches/minecraft/net/minecraft/command/Commands.java.patch (hunk 3 above)
Comment (ichttt) [reply patches/minecraft/net/minecraft/command/Commands.java.patch]: I somewhat understand this, but please add a brief version of this to the code comment so that when someone needs to port patches they understand why this is there as well

**Comment by coehlrich** [comment]
Currently when using this, when connecting to a non forge and running a server command it doesn't get sent to the server and when running a client command the command doesn't get added to the client's sent history.

File: src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java (hunk 1 above)
Comment (GotoLink) [reply src/main/java/net/minecraftforge/client/event/CommandRegistryEvent.java:29]: The RegisterCommandsEvent is there to fix registration originating from vanilla datapacks, and triggers on both client and server side. The registered command won't execute on the client and may or may not execute on the server, depending on various conditions.

While this PR is intended to be a long-term feature supported by Forge, and a new instalment of what was supported in older versions a.k.a _client-only_ 'registered **and executed**' commands, the execution is guaranteed not conditional.
And as you said, they trigger at different times so if a modder used those for a generic listener, he would not get the same behaviour with both event.

**Comment by GotoLink** [comment]
Yes there is a patch in a server side class included in this PR, which exists for the sole purpose of fixing this issue.
The client sent history isn't updated for a client command because the message is never sent to the server when it is processed by the client, as intended.

**Comment by GotoLink** [comment]
I have added more precise comments, proper javadoc in the client hook method.
And another patch to fix what coehlrich mentionned above: in case of vanilla server, clients can strip the empty lambdas on their own.

**Comment by TheCurle** [comment]
If someone has the time to review this, i'd appreciate it.

**Comment by QWERTY770** [comment]
Please remove the lable "Stale".

**Comment by coehlrich** [comment]
I made an updated version of this PR: #7754 
It also makes a couple of changes.

**Comment by sciwhiz12** [comment]
Hello! @GotoLink, please leave a comment whether you wish to continue this PR for 1.16. Note that there is already a 1.18 version of the feature intended by this PR at #7754, so it is your decision whether to discontinue updating this PR in favor of that or to continue with your own implementation. (We suggest the former so there is a concentration of effort in one PR rather than fragmented across multiple.) If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by sciwhiz12** [comment]
Hello! Since the application of the Stale label, there has been no recent activity on this PR. As said in our previous comment, this PR will now be closed for inactivity.

If you wish to continue this PR, please open a new PR against the Latest supported version (to be conformant with our version support and contributions policy).

As a note, PR #7754 which implements a similar feature to this has been merged since the previous comment.

Thank you for your contributions. 👋
